osquery/tables/networking/curl_certificate.cpp

```diff
@@ -335,49 +352,133 @@ Status getTLSCertificate(const std::string& hostname,
     return Status::failure("Failed to create OpenSSL CTX object");
   }
 
-  auto delBIO = [](BIO* bio) { BIO_free_all(bio); };
-  auto server = std::unique_ptr<BIO, decltype(delBIO)>(
-      BIO_new_ssl_connect(ctx.get()), delBIO);
-  if (server == nullptr) {
-    return Status::failure("Failed to create OpenSSL BIO object");
+  std::string port = "443";
+  auto connect_hostname = hostname;
+  auto delim = hostname.find(":");
+  if (delim + 1 == hostname.length()) {
+    // if no port specified use default port
+    connect_hostname = hostname.substr(0, delim);
+  } else if (delim != std::string::npos) {
+    port = hostname.substr(delim + 1, std::string::npos);
+    connect_hostname = hostname.substr(0, delim);
+  }
+
+  int ret;
+#ifdef WIN32
+  // Initialize Winsock
+  WSADATA wsaData;
+  ret = WSAStartup(MAKEWORD(2, 2), &wsaData);
+  if (ret != 0) {
+    return Status::failure("WSAStartup failed with error: " +
+                           std::to_string(ret));
   }
+#endif
 
-  std::string port = ":443";
-  auto ext_hostname = hostname;
-  auto conn_hostname = hostname;
-  auto delim = hostname.find(":");
-  if (delim == std::string::npos) {
-    conn_hostname += ":443";
-  } else {
-    ext_hostname = hostname.substr(0, delim);
+  // connect up the socket first and then pass to BIO
+  struct addrinfo hints, *addrinfo, *current_addr, *preferred_addr;
+  memset(&hints, 0, sizeof(hints));
+  hints.ai_family = PF_UNSPEC;
+  hints.ai_socktype = SOCK_STREAM;
+  hints.ai_flags = AI_NUMERICSERV | AI_ADDRCONFIG;
+
+  ret = getaddrinfo(connect_hostname.c_str(), port.c_str(), &hints, &addrinfo);
+  if (ret != 0) {
+#ifdef WIN32
+    WSACleanup();
+#endif
+    return Status::failure("Unable to resolve hostname");
   }
 
-  auto ret = BIO_set_conn_hostname(server.get(), conn_hostname.c_str());
-  if (ret != 1) {
-    return Status::failure("Failed to set OpenSSL hostname: " +
-                           std::to_string(ret));
+  current_addr = addrinfo;
+  // take the first return as the default, but prefer IPv4 addrs
+  preferred_addr = addrinfo;
+  while (current_addr) {
+    if (current_addr->ai_family == AF_INET) {
+      preferred_addr = current_addr;
+      break;
+    }
+    current_addr = current_addr->ai_next;
   }
 
-  SSL* ssl = nullptr;
-  BIO_get_ssl(server.get(), &ssl);
-  if (ssl == nullptr) {
-    return Status::failure("Failed to retrieve OpenSSL object");
+#ifdef WIN32
+  SOCKET sock;
+#else
+  int sock;
+#endif
+  sock = socket(preferred_addr->ai_family,
```

Comment:
Is there any cleanup for closing this socket after you confirm `>= 0`?

Reply:
very good question.  We can't close the socket until it's connected.  I then moved an unrelated SSL initialization out of the way.  That leaves us with this: We then hand off the socket to BIO and tell it to close it down when it is done (that's what BIO_CLOSE is suppose to do): https://github.com/osquery/osquery/pull/6641/files#diff-850a3567d411d5f1b144c0ac1e59f1c1R447

The only place where this could potentially go wrong is here: https://github.com/osquery/osquery/pull/6641/files#diff-850a3567d411d5f1b144c0ac1e59f1c1R447 and here I'm not totally sure what the behavior of openssl would be in this circumstance if it's unable to create the BIO.  So, I'm thinking that we shouldn't re-close the socket if the BIO_new_socket fails.